### PR TITLE
Implement Error Trait for BcryptError

### PR DESCRIPTION
**Pull request by drusellers**
Questions to be answered:

- Should the error expose the invalid options supplied by the user?
- Is the `.to_string()` call acceptable on line 93 of lib.rs

### CODE REVIEW

File: src/errors.rs [hunk 1]
```diff
@@ -9,8 +10,8 @@ pub type BcryptResult<T> = Result<T, BcryptError>;
 /// passwords
 pub enum BcryptError {
     Io(io::Error),
-    InvalidCost,
-    InvalidPrefix
+    InvalidCost(u32),
+    InvalidPrefix(String),
```
Comment (drusellers) [inline src/errors.rs]: I'm not pleased with this - but that is because of my newness to rust.

File: src/lib.rs [hunk 2]
```diff
@@ -31,9 +31,9 @@ pub static DEFAULT_COST: u32 = 12;
 #[derive(Debug, PartialEq)]
 /// A bcrypt hash result before concatenating
 struct HashParts {
- cost: u32,
- salt: String,
- hash: String,
+    cost: u32,
+    salt: String,
+    hash: String,
```
Comment (drusellers) [inline src/lib.rs]: Ah, drat. `rustfmt` changed this on me. I can revert.

File: src/lib.rs [hunk 3]
```diff
@@ -70,7 +70,7 @@ fn _hash_password(password: &str, cost: u32, salt: &[u8]) -> BcryptResult<HashPa
     Ok(HashParts {
         cost: cost,
         salt: b64::encode(&salt),
-        hash: b64::encode(&output[..23]) // remember to remove the last byte
+        hash: b64::encode(&output[..23]), // remember to remove the last byte
```
Comment (drusellers) [inline src/lib.rs]: `rustfmt`

File: src/lib.rs [hunk 4]
```diff
@@ -80,30 +80,34 @@ fn split_hash(hash: &str) -> BcryptResult<HashParts> {
     let mut parts = HashParts {
         cost: 0,
         salt: "".to_owned(),
-        hash: "".to_owned()
+        hash: "".to_owned(),
```
Comment (drusellers) [inline src/lib.rs]: `rustfmt`

File: src/lib.rs [hunk 5]
```diff
@@ -80,30 +80,34 @@ fn split_hash(hash: &str) -> BcryptResult<HashParts> {
...
             0 => (),
-            1 => match part {
-                "2y" | "2b" | "2a" => (),
-                _ => { return Err(BcryptError::InvalidPrefix); }
-            },
+            1 => {
+                match part {
+                    "2y" | "2b" | "2a" => (),
+                    supplied => {
+                        return Err(BcryptError::InvalidPrefix(supplied.to_string()));
```
Comment (drusellers) [inline src/lib.rs]: If this is something we want to do, I'm not sure this is the best approach to take.

File: src/lib.rs [hunk 6]
```diff
@@ -145,7 +149,7 @@ mod tests {
         let expected = HashParts {
             cost: 12,
             salt: "L6Bc/AlTQHyd9liGgGEZyO".to_owned(),
-            hash: "FLPHNgyxeEPfgYfBCVxJ7JIlwxyVU3u".to_owned()
+            hash: "FLPHNgyxeEPfgYfBCVxJ7JIlwxyVU3u".to_owned(),
```
Comment (drusellers) [inline src/lib.rs]: `rustfmt`

File: src/lib.rs (hunk 2 above)
Comment (Keats) [inline src/lib.rs]: no need, it's fine

File: src/lib.rs (hunk 5 above)
Comment (Keats) [inline src/lib.rs]: it's fine like this

**Pull request merged**